airbyte-integrations/connectors/source-google-analytics-v4/source_google_analytics_v4/custom_reports_validator.py

```diff
@@ -103,6 +104,8 @@ def validate(self):
         try:
             for report in self.reports:
                 self.model.parse_obj(report)
+                if "segments" in report and "ga:segment" not in report["dimensions"]:
+                    raise ValueError("ga:segment is required in dimensions if segments are provided")
```

Comment:
We should extend `class Model(BaseModel):` (method `check_field_reference_forrmat`) for this type of validation.

Reply:
@artem1205 the `@validator` decorator runs each field separated so it's not possible to validate both (maybe I'm wrong) and the `@root_validator(pre=True)` works but because there is a post error parser I don't quite understand why it's capturing the previous error message.

```python
    @root_validator(pre=True)
    def check_segment_included_in_dimension(cls, values):
        dimensions = values.get("dimensions") 
        segments = values.get("segments")
        print(segments, dimensions)
        if segments and "ga:segment" not in dimensions:
            raise ValueError(ERROR_MSG_MISSING_SEGMENT_DIMENSION)
        return values
```

Whatever I run the test the error message is:
```
E           assert 'errors: `ga:segment` is required' in "Custom Reports has invalid structure in report: {'name': 'missing_segment_dimension', 'dimensions': ['ga:test'], 'segments': ['another_segment'], 'metrics': ['ga:test']}, errors: fields not permitted ['added_field']"
E            +  where "Custom Reports has invalid structure in report: {'name': 'missing_segment_dimension', 'dimensions': ['ga:test'], 'segments': ['another_segment'], 'metrics': ['ga:test']}, errors: fields not permitted ['added_field']" = str(AirbyteTracedException("Custom Reports has invalid structure in report: {'name': 'missing_segment_dimension', 'dimensi...: ['ga:test'], 'segments': ['another_segment'], 'metrics': ['ga:test']}, errors: fields not permitted ['added_field']"))
```

As you can see the error message is using from the test before the `missing_segment_dimension`
```python
@pytest.mark.parametrize(
    "custom_reports, expected",
    (
        ([{"name": "test", "dimensions": ["ga+test"], "metrics": ["ga!test"]}], "errors: incorrect field reference"),
        ([{"name": [], "dimensions": ["ga:test"], "metrics": ["ga:test"]}], "errors: type errors"),
        ([{"name": "test", "dimensions": ["ga:test"], "metrics": ["ga:test"], "added_field": "test"}], "errors: fields not permitted"),
        ([{"name": "missing_segment_dimension", "dimensions": ["ga:test"], "segments": ["another_segment"], "metrics": ["ga:test"]}], "errors: `ga:segment` is required"),
        ([{"missing_name": "test", "dimensions": ["ga:test"], "metrics": ["ga:test"]}], "errors: fields required"),
    ),
    ids=["incorrrect field reference", "type_error", "not_permitted", "missing", "missing_segment_dimension"],
)
```

Any recommendation?

Reply:
using root_validator is right approach for this case, to pass parametrized test we should use classmethod:
```python
@classmethod
@root_validator(pre=True)
def check_segment_included_in_dimension(cls, values): ...
```

Reply:
It didn't work. It continue getting the value from the previous test @artem1205 can you help here?